mslib/mscolab/server.py

```diff
@@ -764,28 +766,30 @@ def available_idps():
 @APP.route("/idp_login/", methods=['POST'])
 def idp_login():
     """Handle the login process for the user by selected IDP"""
-    selected_idp = request.form.get('selectedIdentityProvider')
-    sp_config = None
-    for idp_config in setup_saml2_backend.CONFIGURED_IDPS:
-        if selected_idp == idp_config['idp_identity_name']:
-            sp_config = idp_config['idp_data']['saml2client']
-            break
+    if mscolab_settings.USE_SAML2:
+        selected_idp = request.form.get('selectedIdentityProvider')
+        sp_config = None
+        for idp_config in setup_saml2_backend.CONFIGURED_IDPS:
+            if selected_idp == idp_config['idp_identity_name']:
+                sp_config = idp_config['idp_data']['saml2client']
+                break
 
-    try:
-        _, response_binding = sp_config.config.getattr("endpoints", "sp")[
-            "assertion_consumer_service"
-        ][0]
-        entity_id = get_idp_entity_id(selected_idp)
-        _, binding, http_args = sp_config.prepare_for_negotiated_authenticate(
-            entityid=entity_id,
-            response_binding=response_binding,
-        )
-        if binding == BINDING_HTTP_REDIRECT:
-            headers = dict(http_args["headers"])
-            return redirect(str(headers["Location"]), code=303)
-        return Response(http_args["data"], headers=http_args["headers"])
-    except (NameError, AttributeError):
-        return render_template('errors/403.html'), 403
+        try:
+            _, response_binding = sp_config.config.getattr("endpoints", "sp")[
+                "assertion_consumer_service"
+            ][0]
+            entity_id = get_idp_entity_id(selected_idp)
+            _, binding, http_args = sp_config.prepare_for_negotiated_authenticate(
+                entityid=entity_id,
+                response_binding=response_binding,
+            )
+            if binding == BINDING_HTTP_REDIRECT:
+                headers = dict(http_args["headers"])
+                return redirect(str(headers["Location"]), code=303)
+            return Response(http_args["data"], headers=http_args["headers"])
+        except (NameError, AttributeError):
+            return render_template('errors/403.html'), 403
+    return render_template('errors/403.html'), 403
 
 
 @APP.route('/<path:url>', methods=['POST'])
```

Comment:
This "catch-all" route makes it so that any request to a non-existing route with a method other than POST now returns a 405 error instead of a 404. E.g. going to `<mscolab-server>/this-route-does-not-exist` in a browser previously returned 404 Not Found and with this change returns 405 Method Not Allowed. I think we should find a way to change that back to a 404.

Reply:
I am wondering, since this route effectively only handles POST requests to the URL's specified in `idp_config['idp_data']['assertion_consumer_endpoints']` (right?), is it possible to dynamically define the routes for the specified endpoints? I.e. loop over the endpoints and define `@APP.route('/<assertion_consumer_endpoint>/', methods=["POST"])` functions (or something similar) which handle them, completely removing the catch-all route and restoring flasks default behavior of 404's for GET requests to undefined routes.

Reply:
Thank you.. Resolved :)